src/Microsoft.Net.Http.Headers/HeaderUtilities.cs

```diff
@@ -480,6 +480,89 @@ public static unsafe bool TryParseInt64(StringSegment value, out long result)
             }
         }
 
+        // Strict and fast RFC7231 5.3.1 Quality value parser (and without memory allocation)
+        // See https://tools.ietf.org/html/rfc7231#section-5.3.1
+        internal static bool TryParseQualityDouble(string input, int startIndex, out double quality, out int length)
+        {
+            quality = 0;
+            length = 0;
+
+            var inputLength = input.Length;
+            var current = startIndex;
+            var limit = startIndex + _qualityValueMaxCharCount;
+
+            var intPart = 0;
+            var decPart = 0;
+            var decPow = 1;
+
+            if (current >= inputLength)
+            {
+                return false;
+            }
+
+            var ch = input[current];
+
+            if (ch >= '0' && ch <= '1')     // Only values between 0 and 1 are accepted, according to RFC
+            {
+                intPart = ch - '0';
+                current++;
+            }
+            else
+            {
+                // The RFC doesn't allow decimal values starting with dot. I.e. value ".123" is invalid. It must be in the
+                // form "0.123".
+                return false;
+            }
+
+            if (current < inputLength)
+            {
+                ch = input[current];
+
+                if (ch >= '0' && ch <= '9')
+                {
+                    // The RFC accepts only one digit before the dot
+                    return false;
+                }
+
+                if (ch == '.')
+                {
+                    current++;
+
+                    while (current < inputLength)
+                    {
+                        ch = input[current];
+                        if (ch >= '0' && ch <= '9')
```

Comment:
This is still going to run for parsing `1.111` for example. Although this is redundant work since any non zero digit after a `1.` it's not a big cost so it's fine as is. 

However, let's add a check at the end before returning to make sure the value is between 0 and 1 inclusive since this helper method parses quality values explicitly and is not a generic double parser. The bounds checking can then be removed here: https://github.com/aspnet/HttpAbstractions/blob/87cd79d6fc54bb4abf07c1e380cd7a9498a78612/src/Microsoft.Net.Http.Headers/StringWithQualityHeaderValue.cs#L221-L224